exec/stats.c

```diff
@@ -363,17 +379,128 @@ cs_error_t stats_map_get(const char *key_name,
 			cs_ipcs_get_global_stats(&ipcs_global_stats);
 			stats_map_set_value(statinfo, &ipcs_global_stats, value, value_len, type);
 			break;
+		case STAT_SCHEDMISS:
+			if (sscanf(key_name, "stats.schedmiss.%d", &num) != 1) {
+				return CS_ERR_NOT_EXIST;
+			}
+
+			sm_event = num;
+
+			if (schedmiss_event[next_schedmiss_event].timestamp != 0) { /* We've wrapped round */
+				sm_event = (sm_event + next_schedmiss_event) % MAX_SCHEDMISS_EVENTS;
+			}
+
+			sm_type = strrchr(key_name, '.');
+			if (sm_type == NULL) {
+				return CS_ERR_NOT_EXIST;
+			}
+			sm_type++;
+
+			if (sm_event > MAX_SCHEDMISS_EVENTS) {
+				sm_event -= MAX_SCHEDMISS_EVENTS;
+			}
+			if (strcmp(sm_type, "timestamp") == 0) {
+				memcpy(value, &schedmiss_event[sm_event].timestamp, sizeof(uint64_t));
+				*value_len = sizeof(uint64_t);
+				*type = ICMAP_VALUETYPE_UINT64;
+			}
+			if (strcmp(sm_type, "delay") == 0) {
+				memcpy(value, &schedmiss_event[sm_event].delay, sizeof(float));
+				*value_len = sizeof(float);
+				*type = ICMAP_VALUETYPE_FLOAT;
+			}
+			break;
 		default:
 			return CS_ERR_LIBRARY;
 	}
 	return CS_OK;
 }
 
-#define STATS_CLEAR       "stats.clear."
-#define STATS_CLEAR_KNET  "stats.clear.knet"
-#define STATS_CLEAR_IPC   "stats.clear.ipc"
-#define STATS_CLEAR_TOTEM "stats.clear.totem"
-#define STATS_CLEAR_ALL   "stats.clear.all"
+static void schedmiss_clear_stats(void)
+{
+	int i;
+	char param[ICMAP_KEYNAME_MAXLEN];
+
+	for (i=0; i<MAX_SCHEDMISS_EVENTS; i++) {
+		schedmiss_event[i].timestamp = (uint64_t)0LL;
+		schedmiss_event[i].delay = 0.0f;
+
+		sprintf(param, "stats.schedmiss.%i.timestamp", i);
+		stats_rm_entry(param);
+		sprintf(param, "stats.schedmiss.%i.delay", i);
+		stats_rm_entry(param);
+	}
+}
+
+static void trigger_schedmiss_trackers(unsigned int sm_index)
+{
+	struct cs_stats_tracker *tracker;
+	struct qb_list_head *iter;
+	struct icmap_notify_value new_val;
+	struct icmap_notify_value old_val;
+	char param[ICMAP_KEYNAME_MAXLEN];
+
+	qb_list_for_each(iter, &stats_tracker_list_head) {
+
+		tracker = qb_list_entry(iter, struct cs_stats_tracker, list);
+		if ( !(tracker->events & ICMAP_TRACK_PREFIX) ||
+		     strncmp(tracker->key_name, "stats.schedmiss",15) ) {
```

Comment:
Eventho it is quite readable I would rather like to have a `strlen` here - "stats.schedmiss" is used multiple times in source code so maybe candidate for `#define`